pkg/scripts/install.go

```diff
@@ -13,177 +15,234 @@ import (
+		return &api.InstallResult{Script: script, URL: location}
 	}
-
-	// get scripts from source
-	sourceResults = make(map[string]*downloadResult, len(scripts))
-	for _, script := range scripts {
-		sourceResults[script] = &downloadResult{location: api.SourceScripts}
-		file := filepath.Join(dstDir, api.SourceScripts, script)
-		if !i.fs.Exists(file) {
-			sourceResults[script].err = errors.NewDownloadError(file, -1)
-		}
-		checkLocation := filepath.Join(dstDir, strings.TrimSuffix(api.SourceScripts, "/bin"), script)
-		if i.fs.Exists(checkLocation) {
-			glog.Warningf("The %q script found in .sti/%s, should be in .sti/bin/%s.", script, script, script)
-		}
+	location = strings.Replace(location, "s2i", "sti", 1)
+	if s.fs.Exists(location) {
+		glog.Infof("DEPRECATED: Use .s2i/bin instead of .sti/bin")
+		return &api.InstallResult{Script: script, URL: location}
 	}
+	return nil
+}
 
-	// get scripts from default URL
-	defaultURL, err := i.docker.GetScriptsURL(i.image)
-	if err == nil && defaultURL != "" {
-		defaultResults = i.download(defaultURL, scripts, filepath.Join(dstDir, api.DefaultScripts))
+// Install copies the script into upload directory and fix its permissions.
+func (s *SourceScriptHandler) Install(r *api.InstallResult) error {
+	dst := filepath.Join(s.DestinationDir, api.UploadScripts, r.Script)
+	if err := s.fs.Rename(r.URL, dst); err != nil {
+		return err
+	}
+	if err := s.fs.Chmod(dst, 0755); err != nil {
+		return err
+	}
+	// Make the path to scripts nicer in logs
+	parts := strings.Split(r.URL, "/")
+	if len(parts) > 3 {
+		r.URL = "<source-dir>/" + strings.Join(parts[len(parts)-3:], "/")
 	}
+	r.Installed = true
+	r.Downloaded = true
+	return nil
+}
 
-	return i.install(scripts, userResults, sourceResults, defaultResults, dstDir)
+// SetDestinationDir sets the directory where the scripts should be uploaded.
+// In case of SourceScriptHandler this is a source directory root.
+func (s *SourceScriptHandler) SetDestinationDir(baseDir string) {
+	s.DestinationDir = filepath.Join(baseDir)
 }
 
-func (i *installer) download(scriptsURL string, scripts []string, dstDir string) map[string]*downloadResult {
-	result := make(map[string]*downloadResult, len(scripts))
+// ScriptSourceManager manages various script handlers.
+type ScriptSourceManager interface {
+	Add(ScriptHandler)
+	SetDownloader(Downloader)
+	Installer
+}
 
-	for _, script := range scripts {
-		result[script] = &downloadResult{location: scriptsURL}
-		url, err := url.Parse(scriptsURL + "/" + script)
-		if err != nil {
-			result[script].err = err
-			continue
-		}
-		_, result[script].err = i.downloader.Download(url, filepath.Join(dstDir, script))
-	}
+// DefaultScriptSourceManager manages the default script lookup and installation
+// for source-to-image.
+type DefaultScriptSourceManager struct {
+	Image      string
+	ScriptsURL string
+	download   Downloader
+	docker     docker.Docker
+	dockerAuth dockerClient.AuthConfiguration
+	sources    *list.List
```

Comment:
@soltysh the array will be probably good here as well, I prefer List because it guarantees the order and it is not that much performance hit. But yeah, I will switch it to use array.